src/protobuf-net.AspNetCore/Formatters/ProtoInputFormatter.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Microsoft.AspNetCore.Mvc.Formatters;
+using ProtoBuf.Meta;
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading.Tasks;
+
+namespace ProtoBuf.Formatters
+{
+    /// <summary>
+    /// Implements a protobuf-net based input formatter
+    /// </summary>
+    public sealed class ProtoInputFormatter : InputFormatter
+    {
+        private readonly TypeModel _model;
+
+        /// <summary>
+        /// Create a new <see cref="ProtoInputFormatter"/> instance
+        /// </summary>
+        /// <param name="model">The type-model to use for deserialization</param>
+        public ProtoInputFormatter(TypeModel model = null)
+        {
+            _model = model ?? RuntimeTypeModel.Default;
+        }
+
+        /// <inheritdoc/>
+        protected override bool CanReadType(Type type)
+            => _model.CanSerializeContractType(type);
+
+        /// <inheritdoc/>
+        public override Task<InputFormatterResult> ReadAsync(InputFormatterContext context)
+        {
+            // default implementation does length checks; we *really* don't want those, since
+            // zero length is a perfectly valid payload in protobuf
+            return ReadRequestBodyAsync(context);
+        }
+
+        /// <inheritdoc/>
+        public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context)
+        {
+            // note: *today*, protobuf-net lacks fully async read, so this buffers in the pipe until
+            // we get the Content-Length that was advertised by the caller
+
+            if (context is null) throw new ArgumentNullException(nameof(context));
+            var request = context.HttpContext.Request;
+            var length = request.ContentLength ?? -1;
```

Comment:
No chunked support? The JSON.NET implementation using input buffering here to read the entire stream to either memory or disk based on a threshold.

Reply:
I mean, sure we could do the same

Reply:
We have a built in type for that. Here's what that looks like: https://github.com/dotnet/aspnetcore/blob/fb671009ebb91f088f55c17dca2bcbd4378544f8/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L128-L162

Not the most efficient thing but maybe reasonable over some size threshold.

Reply:
@davidfowl  added chunked plus oversized support (threshold on options) - see `StreamBufferedAsync`